src/crate/client/sqlalchemy/tests/dict_test.py

```diff
@@ -89,18 +90,19 @@ def test_select_with_dict_column_where_clause_other_col(self):
         s = s.where(mytable.c.data['x'] == mytable.c.name)
         self.assertSQL(
             "SELECT mytable.name FROM mytable " +
-            "WHERE mytable.data[:data_1] = mytable.name",
+            "WHERE mytable.data['x'] = mytable.name",
             s
         )
 
     def test_update_with_dict_column(self):
         mytable = self.mytable
         stmt = mytable.update().\
             where(mytable.c.name == 'Arthur Dent').\
-            values({mytable.c.data['x']: "Trillian"})
-
+            values({
+                "data['x']": "Trillian"
+            })
```

Comment:
@hammerhead: Oh, I forgot to keep this one when reverting your change. Apologies. Maybe you can add it to GH-485 again, with a descriptive commit message, which outlines that there was an actual programming error here?

Reply:
Done: 1429658ec9001928c3a389618e2a4ea2b25b8ad1

I checked it a bit more in detail, and found that `parameters` in `Cursor.execute` differ: Without my change, it receives a `BindParameter` (`(BindParameter('%(4541705392 param)s', 'Trillian', type_=NullType()), '1')`). With my change, it receives a `Tuple` (`('Trillian', '1')`).

Both variants seem to carry the same information, but the current implementation fails to handle the `BindParameter` correctly when attempting to serialize it to JSON for the HTTP call. So it might also be considered a missing feature that what the test was previously trying to do would fail (potentially used to work earlier?).

Reply:
The fix is rather trivial:
```diff
diff --git a/src/crate/client/http.py b/src/crate/client/http.py
index e932f73..ac03ae8 100644
--- a/src/crate/client/http.py
+++ b/src/crate/client/http.py
@@ -52,6 +52,7 @@ from crate.client.exceptions import (
     DigestNotFoundException,
     ProgrammingError,
 )
+from sqlalchemy.sql.expression import BindParameter
 
 
 logger = logging.getLogger(__name__)
@@ -93,6 +94,9 @@ class CrateJsonEncoder(json.JSONEncoder):
                        (delta.seconds + delta.days * 24 * 3600) * 1000.0)
         if isinstance(o, date):
             return calendar.timegm(o.timetuple()) * 1000
+        if isinstance(o, BindParameter):
+            return o.value
+
         return json.JSONEncoder.default(self, o)
```
But I was unable to produce a test case yet that would expose the issue or generally asses how much of a problem this is.

Reply:
Hi Niklas,

this is intriguing. Thank you for sharing your findings. Based on this information, I am thinking about if 1429658e should better be handed in separately, maybe together with the other fix you shared above, and two _integration test_ cases (without using mocking), which exercise and verify that both variants work.

> But I was unable to produce a test case yet that would expose the issue.

GH-491 may come to the rescue for finding an appropriate place for those integration tests, so this is probably the right chance to spawn it. I will stage a corresponding PR where your fixes can then be picked into.

With kind regards,
Andreas.